### PR TITLE
Allow CONCAT(varchar(x), ...) to return limited varchar type

### PR DESCRIPTION
**Pull request by fiedukow**

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ConcatFunction.java [hunk 1]
```diff
@@ -163,4 +181,89 @@ public static int checkedAdd(int x, int y)
...
+        {
+            ImmutableList.Builder<TypeSignature> boundTypesSignatures = ImmutableList.builder();
+
+            long lengthAccumulator = 0;
+            for (Type type : actualArgumentTypes) {
+                if (!isTypeValid(type)) {
+                    return Optional.empty();
+                }
+
+                int paramLength = varcharLength(type);
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ConcatFunction.java]: You do not take coercions into account in `varcharLength`. And `isTypeValid` allows ending up here with type being anything coercible to `varchar`. It seems not right to me.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ConcatFunction.java [hunk 2]
```diff
@@ -163,4 +181,89 @@ public static int checkedAdd(int x, int y)
...
+                            declaredSignature.getName(),
+                            declaredSignature.getKind(),
+                            VarcharType.createVarcharType(Ints.saturatedCast(lengthAccumulator)).getTypeSignature(),
+                            boundTypesSignatures.build()));
+        }
+
+        @Override
+        public Optional<BoundVariables> bindVariables(List<? extends Type> actualArgumentTypes, Type actualReturnType)
+        {
+            Optional<Signature> signature = bind(actualArgumentTypes);
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ConcatFunction.java]: Why return any map with bound variables while input signature have no variables at all.
You just put values for `x<num>` and `r` - and those will never be used later on anyway.
Right?

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ConcatFunction.java [hunk 3]
```diff
@@ -163,4 +181,89 @@ public static int checkedAdd(int x, int y)
...
+            if (!signature.isPresent()) {
+                return Optional.empty();
+            }
+
+            if (!isTypeValid(actualReturnType)) {
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ConcatFunction.java]: how is that possible?

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ConcatFunction.java [hunk 4]
```diff
@@ -163,4 +181,89 @@ public static int checkedAdd(int x, int y)
...
+                return Optional.empty();
+            }
+
+            int returnLength = varcharLength(actualReturnType);
+            int calculatedReturnLength = Ints.saturatedCast(getOnlyElement(signature.get().getReturnType().getParameters().iterator()).getLongLiteral());
+            if (returnLength < calculatedReturnLength) {
+                return Optional.empty();
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ConcatFunction.java]: Also here?
I think we need to discuss this offline.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ConcatFunction.java [hunk 5]
```diff
@@ -163,4 +181,89 @@ public static int checkedAdd(int x, int y)
...
+            if (allowCoercion && typeManager.coerceTypeBase(type, StandardTypes.VARCHAR).isPresent()) {
+                return true;
+            }
+
+            if (type.getTypeSignature().getBase() != StandardTypes.VARCHAR) {
+                return false;
+            }
+
+            TypeSignatureParameter param = getOnlyElement(type.getTypeSignature().getParameters().iterator());
+            checkArgument(param.getKind() == ParameterKind.LONG);
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ConcatFunction.java]: isn't this check redundant? Parameters should be verified when `VarcharType` is created.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ConcatFunction.java [hunk 6]
```diff
@@ -163,4 +181,89 @@ public static int checkedAdd(int x, int y)
...
+            ImmutableList.Builder<TypeSignature> boundTypesSignatures = ImmutableList.builder();
+
+            long lengthAccumulator = 0;
+            for (Type type : actualArgumentTypes) {
+                if (!isTypeValid(type)) {
+                    return Optional.empty();
+                }
+
+                int paramLength = varcharLength(type);
+                lengthAccumulator = Ints.saturatedCast(lengthAccumulator + paramLength);
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ConcatFunction.java]: why saturated casts is required?

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ConcatFunction.java (hunk 2 above)
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ConcatFunction.java]: +1, those `x` and `r` don't seem to be used

File: presto-main/src/test/java/com/facebook/presto/sql/TestExpressionInterpreter.java [hunk 7]
```diff
@@ -992,7 +992,7 @@ public void testTry()
         assertOptimizedEquals("TRY(ABS(-2))", "2");
         assertOptimizedEquals("42 + TRY(ABS(-9223372036854775807 - 1))", "null");
 
-        assertOptimizedEquals("JSON_FORMAT(TRY(JSON '[]')) || unbound_string", "'[]' || unbound_string");
+        assertOptimizedEquals("JSON_FORMAT(TRY(JSON '[]')) || unbound_string", "CAST('[]' AS VARCHAR) || unbound_string");
```
Comment (sopel39) [inline presto-main/src/test/java/com/facebook/presto/sql/TestExpressionInterpreter.java]: is this change required?

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ConcatFunction.java (hunk 5 above)
Comment (fiedukow) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ConcatFunction.java]: All sanity checks are redundant ;-P

File: presto-main/src/main/java/com/facebook/presto/metadata/DefaultSignatureBinder.java [hunk 8]
```diff
@@ -0,0 +1,449 @@
...
+
+        calculateVariableValuesForLongConstraints(variableBinder);
+        BoundVariables boundVariables = variableBinder.build();
+        if (!allTypeVariablesBound(boundVariables)) {
+            return Optional.empty();
+        }
+        return Optional.of(boundVariables);
+    }
+
+    public static Signature bindVariables(Signature signature, BoundVariables boundVariables, int arity)
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/metadata/DefaultSignatureBinder.java]: It's strange to have both static and non-static methods (also with different return value) with the same name in the same class. It suggest that those methods do the same logically.

I would suggest to refactor `bindVariables` that do not take `actualReturnType` into:
`Optional<BoundVariables> bindArgumentVariables(...)` or similar. You could then use those methods in `bindVariables(List<? extends Type> actualArgumentTypes, Type actualReturnType)`.

Static methods should go to the top of class.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ConcatFunction.java (hunk 6 above)
Comment (fiedukow) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ConcatFunction.java]: Becasue max length of VARCHAR is MAX_INT and two ints may have bigger value. In such case we just want to retrurn VARCHAR(MAX_INT).

Typical case for that is when one of CONCAT arguments has length of MAX_INT as it is result of other calculation not allowing to limit VARCHAR size to anything smaller.

File: presto-main/src/test/java/com/facebook/presto/sql/TestExpressionInterpreter.java (hunk 7 above)
Comment (fiedukow) [inline presto-main/src/test/java/com/facebook/presto/sql/TestExpressionInterpreter.java]: yes

File: presto-main/src/main/java/com/facebook/presto/metadata/DefaultSignatureBinder.java [hunk 9]
```diff
@@ -0,0 +1,449 @@
...
+            builder.addAll(argumentSignatures);
+            for (int i = 0; i < variableArityArgumentsCount; i++) {
+                builder.add(getLast(argumentSignatures));
+            }
+            return builder.build();
+        }
+        return argumentSignatures;
+    }
+
+    public static List<TypeSignature> bindVariables(List<TypeSignature> typeSignatures, BoundVariables boundVariables)
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/metadata/DefaultSignatureBinder.java]: Please give more meaningful name. Method `bindVariables` takes `boundVariables` as argument? Doesn't feel right.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ConcatFunction.java (hunk 6 above)
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ConcatFunction.java]: Still, you should you use VarcharType.MAX_INT value explicitly to be transparent about Varchar length limit.

File: presto-main/src/main/java/com/facebook/presto/metadata/FunctionRegistry.java [hunk 10]
```diff
@@ -969,7 +969,7 @@ public static OperatorType unmangleOperator(String mangledName)
     public boolean isMoreSpecificThan(ApplicableFunction left, ApplicableFunction right)
     {
         List<Type> resolvedTypes = resolveTypes(left.getBoundSignature().getArgumentTypes(), typeManager);
-        Optional<BoundVariables> boundVariables = new SignatureBinder(typeManager, right.getDeclaredSignature(), true)
+        Optional<BoundVariables> boundVariables = new DefaultSignatureBinder(typeManager, right.getDeclaredSignature(), true)
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/metadata/FunctionRegistry.java]: I think you should use `signatureBinder` from the right `ApplicableFunction`. Maybe `ApplicableFunction` should contain `SqlFunction` or `signatureBinder` field in addition to `declaredSignature`.

File: presto-main/src/test/java/com/facebook/presto/metadata/TestSignatureBinder.java [hunk 11]
```diff
@@ -29,12 +29,12 @@
 import java.util.Optional;
```
Comment (sopel39) [inline presto-main/src/test/java/com/facebook/presto/metadata/TestSignatureBinder.java]: rename to `TestDefaultSignatureBinder`

File: presto-main/src/main/java/com/facebook/presto/metadata/DefaultSignatureBinder.java [hunk 12]
```diff
@@ -0,0 +1,449 @@
...
+    {
+        ImmutableList.Builder<TypeSignature> builder = ImmutableList.builder();
+        for (TypeSignature typeSignature : typeSignatures) {
+            builder.add(bindVariables(typeSignature, boundVariables));
+        }
+        return builder.build();
+    }
+
+    public static TypeSignature bindVariables(
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/metadata/DefaultSignatureBinder.java]: ditto

File: presto-main/src/main/java/com/facebook/presto/metadata/DefaultSignatureBinder.java [hunk 13]
```diff
@@ -0,0 +1,449 @@
...
+        }
+
+        List<TypeSignatureParameter> parameters = typeSignature.getParameters().stream()
+                .map(typeSignatureParameter -> bindVariables(typeSignatureParameter, boundVariables))
+                .collect(toList());
+
+        return new TypeSignature(baseType, parameters);
+    }
+
+    private static TypeSignatureParameter bindVariables(
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/metadata/DefaultSignatureBinder.java]: ditto

**Comment by cberner** [comment]
I don't think this is the right approach. `SignatureBinder` implements the rules for how type parameters bind to types, and whether a given function invocation resolves to a specific function signature. The implementation of a function should not be able to influence this process (beyond declaring its signature). I seem to recall there being some work on "type calculations" where you could declare a function signature like: `concat(varchar(x), varchar(y)) -> varchar(x + y)`. Is that something Teradata was working on

**Comment by fiedukow** [comment]
@cberner 
`concat(varchar(x), varchar(y)) -> varchar(x + y)`  is something you can do right now, but it does not apply to variadic functions. 
We've started that work as an idea of how to solve this and didn't have a chance to come back and finish it. The problem with variadic function is that they only allow to declare function having all "variadic" arguments of same type. Note, that varchar(x) and varchar(y) is not the same type. It seems that there is no good way of expressing variadic function. We saw following possibilities:
 1) To use "generic" type as variadic argument allowing function to throw an exception is invalid. This is bad due to late throw and possible ambiguities.
 2) To use list of types that may be binded as function argument. This has same downsides as (1), only less visible.
 3) To use some complex expression for defining which types are allowed. This is hard, long and seems like total overkill.
 4) Only extend mechanism to allow to declare parametric types as an input and bind it with parameters. This has a downside of only supporting single case and more important - it doesn't solve return type calculation. I think we should create interface allowing to calculate function to provide a way to calculate output type based on binded arguments. This make it something similar to PR solution, only a little less invasive and function only gets well defined possibilities.
 5) Allow to override SignatureBinder leaving a choice how to solve the problem to person implementing such function.
 6) Define `n` non variadic variants of CONCAT function for 1,2,3...n arguments using existing type calculation. For for number of arguments greater than n, the CONCAT would still return VARCHAR instead of VARCHAR(x). This seems wrong to me as it is very inconsistent.

If this is important for you we can try to find time to go back to this and implement it, but still it would be nice if you'd give us a clue which solution do you like best, or if you see other solution - show us a way.

**Comment by cberner** [comment]
Solving https://github.com/prestodb/presto/issues/5671 is important for us, as it seems the `varchar(x)` work introduced some backwards incompatibilities. I spent a little while looking at it though, and it seems like it's more an issue of coercions and doesn't require this PR, necessarily.

As to this PR, is there any relevant literature on handling var-args functions with varying types? (3) seems ok to me, if it fits into the `TypeCalculation` framework well, but I agree it may be a lot of work. Another solution might be to rewrite var-args functions like this as recursive invocations, but that would also be a lot of work. I don't like the idea of introducing user defined Java code into the type system though, as that sounds like it will make inference difficult/impossible.

**Comment by fiedukow** [comment]
You're right about this, it would limit UDF possibilities in the future or complicate them dramatically.
As for this task, we currently don't have resources to continue work on that. We will go back to that in future, hopefully near future.

As for #5671 I will investigate that ASAP.

**Comment by fiedukow** [comment]
Rejected in current form.
